scylla-cdc-printer/src/printer.rs

```diff
@@ -0,0 +1,246 @@
+
+use scylla_cdc::cdc_types::GenerationTimestamp;
+use scylla_cdc::reader::StreamReader;
+use scylla_cdc::stream_generations::GenerationFetcher;
+
+struct CDCLogPrinter {
+    session: Arc<Session>,
+    start_timestamp: Duration,
+    end_timestamp: Duration,
+    sleep_interval: time::Duration,
+    window_size: Duration,
+    safety_interval: Duration,
+    readers: Vec<Arc<StreamReader>>,
+}
+
+impl CDCLogPrinter {
+    fn new(
+        session: &Arc<Session>,
+        start_timestamp: Duration,
+        end_timestamp: Duration,
+        window_size: Duration,
+        safety_interval: Duration,
+        sleep_interval: time::Duration,
+    ) -> CDCLogPrinter {
+        CDCLogPrinter {
+            session: Arc::clone(session),
+            start_timestamp,
+            end_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+            readers: vec![],
+        }
+    }
+
+    pub async fn run(
+        &mut self,
+        keyspace: String,
+        table_name: String,
+        mut after_fetch_callback: impl FnMut(Row),
+    ) -> anyhow::Result<()> {
+        let fetcher = &GenerationFetcher::new(&self.session);
+        let mut generation = self.get_first_generation(fetcher).await?;
+
+        loop {
+            self.readers = vec![];
+            for stream_id in fetcher
+                .fetch_stream_ids(&generation)
+                .await?
+                .iter()
+                .flatten()
+            {
+                let reader = StreamReader::new(
+                    &self.session,
+                    vec![stream_id.clone()],
+                    self.start_timestamp,
+                    self.window_size,
+                    self.safety_interval,
+                    self.sleep_interval,
+                );
+                self.readers.push(Arc::new(reader));
+            }
+
+            let handles = self.readers.iter().map(|reader| {
+                tokio::spawn(async move {
+                    reader
+                        .fetch_cdc(keyspace.clone(), table_name.clone(), after_fetch_callback)
+                        .await
+                })
+            });
+
+            generation = loop {
+                match fetcher.fetch_next_generation(&generation).await? {
+                    Some(generation) => break generation,
+                    None => {
+                        let now = SystemTime::now()
+                            .duration_since(SystemTime::UNIX_EPOCH)
+                            .unwrap();
+                        if now + self.sleep_interval >= self.end_timestamp.to_std()? {
+                            break GenerationTimestamp { timestamp: self.end_timestamp };
+                        }
+                        sleep(self.sleep_interval).await
+                    }
+                }
+            };
+
+            self.set_upper_timestamp(generation.timestamp).await;
+
+            futures::future::try_join_all(handles).await?;
```

Comment:
What is `consume_cdc` you mention?

How the errors should be handled?

Reply:
`consume_cdc` is a method of the `Consumer` trait. You should create a separate consumer for each tokio task and feed it CDC rows using the `consume_cdc` method.

Ideally, in case of an error the whole library should stop (gracefully) and return the error.